[Overhaul][1.17.1][Fluids] - Fluid Overhaul PR

Lanse505 opened the pull request:
This PR aims to fix the fundamentally hardcoded nature of Fluids and Fluid Behaviour in 1.16 onwards.

In this PR we've greatly expanded upon the feature set of Custom Fluids to make their behaviour far more configurable as well as expanding upon their control over expected fluid behaviours such as hydrating certain blocks, allowing for more granular control over drowning, extinguishing, etc.

Due to the nature of Fluidlogging in minecraft a lot of what would be considered "block" tied behaviour and not direct "fluid-tied" behaviour needs to be part of the fluid, to allow for fluidloggable blocks that would support certain fluids to actually be able to call and use the behaviour of those fluids. 

With this in mind, in the future ``FluidAttributes`` will have to play a far bigger role in terms of defining certain behaviours of fluids. With this, we've introduced a whole new line of methods and fields to FluidAttributes for people to work with:

The following fields has been added:

Behaviour Relevant Fields:
```  
  - canSwim // Used to dictate whether a player can enter the "Swimming" pose and thus "sprinting" state while inside of the fluid, and thus giving them a speed-boost.
  
  - canDrown // Used to dictate whether an LivingEntity can drown while submerged in the Fluid.
  
  - canExtinguish // Used to dictate whether an entity will be extinguished if touching the fluid and is on fire.
  
  - canHydrate // Used to dictate if this block can provide "hydration" to various blocks. With this ConcretePowderBlock, CoralBlock, FarmBlock, SpongeBlock, and SugarCaneBlock have been patched to support this.
  
  - canBoat // Used to dictate whether a boat can both be placed in the Fluid as well as function within normal behaviour similar to water.
```

Fluid Movement-related Fields:
```  
  - horizontalMotionModifier // Used to dictate the horizontal movement modifier when moving inside of the Fluid.
  
  - enchantmentModifier // Used to dictate an additional horizontal movement modifier. By default this also requires you to override IForgeFluid#handleMotion since the default code is a reimplementation of Water fluid physics which only passes in the "Depth Strider" enchantment for checking.
  
  - effectModifier // Used to dictate a final modifier upon horizontal movement. By default this also requires you to override IForgeFluid#handleMotion since the default code is a reimplementation of Water fluid physics which only passes in the "Dolphin's Grace" effect for checking.
  ```
  
  Misc Fields:
  ```
    - motionScale // Used to scale the motion being applied when a flowing fluid is attempting to push you.
  
    - fallDistanceModifier // Applies as a multiplicative addition onto an entity when it falls through this liquid.
  ```
  
  Most if not all of these exists as well as methods in IForgeFluid and IForgeFluidState which can be overriden if so is wanted.
  As well as allowing for more "granular" control over certain behaviours.
  
  With this PR a few "non-fluid" related changes has been done:
  - BlockPathTypes is now an extendable Enum to allow people to create new BlockPathTypes for pathfinding purposes for their fluids. This is to once again help encourage more granular control over your fluids with the new ``IForgeFluid#getBlockPathType`` and ``IForgeFluid#getAdjacentBlockPathType``
  
  - ``IFluidRenderProperties`` has been introduced similar to ``IItemRenderProperties`` or ``IBlockRenderProperties``
  

Outside of these changes a few notable changes are:
- "isGaseous" is no longer a FluidAttribute field but is now rather a fluid tag that you apply to your fluids.
  - This was done since it had no direct implications on either upwards flowing fluids or upside-down buckets.
  - A PR will be opened adressing upwards-flowing fluids and upside-down buckets in the future.
    - These will continue using the ``FluidAttributes#isLighterThanAir`` method check!
    
- The supplier constructor as well as the ``field_to_method`` transformer for LiquidBlock has been fixed since it was improperly implemented during porting which caused a crash due to not properly redirecting calls from ``LiquidBlock#fluid`` to ``LiquidBlock#getFluid()``

- And finally ``IForgeFluid#isEntityInside``, ``IForgeFluid#isAABBInsideMaterial`` and ``IForgeFluid#isAABBInsideLiquid`` have been removed since there is now dedicated methods on the Entity both as part of IForgeEntity and patches to Entity to get information about the fluid the entity is inside of as well as checks for eye levels inside of the fluid and such.

pupnewfster commented on patches/minecraft/net/minecraft/client/renderer/FogRenderer.java.patch:
```diff
@@ -42,6 +78,16 @@
+ 
+          RenderSystem.m_157445_(-8.0F);
+          RenderSystem.m_157443_(f * 0.5F);
+-      } else {
++      } else if (p_109025_.m_90593_() && !fluidState.m_76152_().getRegistryName().m_135827_().equals("minecraft") && p_109025_.m_90583_().f_82480_ < (double)((float)p_109025_.m_90588_().m_123342_() + fluidState.m_76155_(entity.f_19853_, p_109025_.m_90588_()))) net.minecraftforge.client.RenderProperties.get(fluidState).setupFog(p_109025_, p_109026_, p_109027_);
++      else {
```
This would be more readable and not have any lines removed as part of this patch if it was:
```
} else if (thing) {
body
} else {
```

pupnewfster commented on patches/minecraft/net/minecraft/client/renderer/ScreenEffectRenderer.java.patch:
```diff
@@ -18,7 +18,8 @@
 +            if (!net.minecraftforge.event.ForgeEventFactory.renderWaterOverlay(player, p_110720_))
              m_110725_(p_110719_, p_110720_);
-          }
+-         }
++         } else if (p_110719_.f_91074_.areEyesInCustomFluid()) net.minecraftforge.client.RenderProperties.get(p_110719_.f_91074_.getTouchingFluid()).renderOverlay(p_110719_, p_110720_);
```
if you put the elseif on the next line this would be a slightly smaller patch:
```
}
else if....
```

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/Entity.java.patch:
```diff
@@ -371,6 +415,24 @@
 +
++   protected FluidState touchingFluid = net.minecraft.world.level.material.Fluids.f_76191_.m_76145_();
++   protected Object2DoubleMap<FluidState> customFluidHeight = new it.unimi.dsi.fastutil.objects.Object2DoubleOpenHashMap<>();
++   public void setInFluid(FluidState touchingFluid) {
++       this.touchingFluid = touchingFluid;
++   }
++   public FluidState getTouchingFluid() {
++       return this.touchingFluid;
++   }
++   public boolean isInFluid() {
++       return this.getTouchingFluid() != net.minecraft.world.level.material.Fluids.f_76191_.m_76145_() && this.getFluidHeight(this.getTouchingFluid()) > 0.0D;
++   }
++   public void addFluidHeight(FluidState state, double fluidHeight) {
++       this.customFluidHeight.put(state, fluidHeight);
++   }
++   public double getFluidHeight(FluidState state) {
++       return this.customFluidHeight.getDouble(state);
++   }
```
I feel like it may make sense to put a reference of each of these methods in `IForgeEntity` so that javadocs can be added saying when to use them, especially for `setInFluid` as it almost feels like something that shouldn't be external but has to be because of how things are hardcoded in entities so it would be good for modders to know when to call it.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/Entity.java.patch:
```diff
@@ -371,6 +415,24 @@
++   public FluidState getTouchingFluid() {
++       return this.touchingFluid;
++   }
++   public boolean isInFluid() {
++       return this.getTouchingFluid() != net.minecraft.world.level.material.Fluids.f_76191_.m_76145_() && this.getFluidHeight(this.getTouchingFluid()) > 0.0D;
++   }
```
I believe this method can be removed from here as it is already being added to `IForgeEntity` with the same body.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -392,6 +417,29 @@
+          double d7;
+          if (this.m_20077_()) {
+             d7 = this.m_20120_(FluidTags.f_13132_);
+-         } else {
++         } else if (this.isInFluid()) d7 = this.getFluidHeight(this.getTouchingFluid());
++         else {
```
Again the patch could be slightly smaller if it was:
```
} else if (thing) {
body
} else {
```
or maybe given the extra line doesn't really gain much in readability here.
```
} else if (thing) { body
} else {
```

LexManos replied:
No Pup, the body of elses should not be on the same line. The rules about patch sizes are for clarity and simplicity. Not a hard line number. He should move the body to a new line, and read the braces.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/vehicle/Boat.java.patch:
```diff
@@ -1,5 +1,14 @@
+                blockpos$mutableblockpos.m_122178_(l1, k1, i2);
+                FluidState fluidstate = this.f_19853_.m_6425_(blockpos$mutableblockpos);
+-               if (fluidstate.m_76153_(FluidTags.f_13131_)) {
++               if (fluidstate.m_76153_(FluidTags.f_13131_) || fluidstate.canBoat(this)) {
```
Given this is having to be changed anyways assuming that `canBoat` defaults to the isIn(Tag Water) check, it would be more readable if these patches just got changed to `if (fluidstate.canBoat(this)) {`. From a brief glance it looks like this isn't the default, though it should be made the default so any mods that call `canBoat` on the fluidstate get the correct result instead of also having to check if it is in the tag. That way modders can also define if they want a custom type of water that behaves as water for all purposes and is H2O but doesn't let boats float in it.

pupnewfster commented on patches/minecraft/net/minecraft/world/level/block/LiquidBlock.java.patch:
```diff
@@ -52,6 +58,12 @@
                 return false;
              }
+          }
+-      }
++      } else if (!this.f_54689_.m_76108_(FluidTags.f_13131_)) return p_54699_.m_60819_().handleFluidInteraction(p_54697_, p_54698_);
```
Putting the else if on the next line will reduce the patch size a bit
```
}
else if thing
```

KingLemming commented:
I haven't reviewed this extensively but have been in the channels where this was being discussed. From a design point of view, it's all sounded fantastic.

Will go over it sometime this week.

pupnewfster reviewed (commented):
I only started skimming through it so I probably will have more comments later on when I continue to look at it, but added some comments as a starter.

The big thing I noticed is that some of the default values for things seem to not fully match vanilla as the vanilla check is left in place with an or forge method. The two main issues I see with this are:
1. If modders want to call those methods generically they may not always return the correct values for the vanilla fluids
2. I think the vanilla compat specifically with forge client to vanilla server is slightly broken for it in that if water/lava gets removed from the corresponding tags for some reason because the server owner is insane, the client may have some desync issues in what it is trying to do in relation to things (though maybe because of how things are done right now it would fallback fine and just have the same issue as my first point in relation to things like if lava gets added to the water tag in theory players should be able to swim in it but it would be returning false from canSwim)

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeLivingEntity.java:
```diff
@@ -0,0 +1,23 @@
+
+public interface IForgeLivingEntity {
+
+    private LivingEntity getLivingEntity()
```
This method should be renamed to `self` to be in line with the other forge extensions.

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeFlowingFluid.java:
```diff
@@ -0,0 +1,44 @@
+
+public interface IForgeFlowingFluid extends IForgeFluid
+{
+    default FlowingFluid self()
```
This should be made private to match forge's other extension interfaces.

pupnewfster commented on src/main/java/net/minecraftforge/common/data/ForgeFluidTagsProvider.java:
```diff
@@ -22,6 +22,7 @@
 import net.minecraftforge.common.ForgeMod;
+import net.minecraftforge.common.Tags.Fluids;
 
 import static net.minecraftforge.common.Tags.Fluids.MILK;
```
This import should just change to be similar to forge's block/item tag providers in being a wildcard import and then remove the need to import be adding the `Tags.Fluids` import above and keep the addTags references a bit cleaner.

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -884,7 +884,7 @@ public static FluidAttributes createVanillaFluidAttributes(Fluid fluid)
                     new ResourceLocation("block/lava_flow"))
                     .translationKey("block.minecraft.lava")
-                    .luminosity(15).density(3000).viscosity(6000).temperature(1300)
+                    .luminosity(15).density(3000).viscosity(6000).temperature(1300).canSwim(a -> false)
```
I think this `canSwim(a -> false)` can be removed now that the changes to there being proper defaults have been done.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/vehicle/Boat.java.patch:
```diff
@@ -9,6 +18,51 @@
+ 
+    protected boolean m_7310_(Entity p_38390_) {
+-      return this.m_20197_().size() < 2 && !this.m_19941_(FluidTags.f_13131_);
++      return this.m_20197_().size() < 2 && !(this.areEyesInCustomFluid() && this.getTouchingFluid().canBoat(this));
```
I am not sure if the naming of the method is just bad but this is changing it from: not in water to not in custom fluid or can't boat. The can't boat part seems correct, the not in water -> not in custom fluid seems a bit odd unless the fact that it has custom in the method name isn't quite right.

pupnewfster commented on patches/minecraft/net/minecraft/world/level/block/ConcretePowderBlock.java.patch:
```diff
@@ -0,0 +1,39 @@
++      return touchesLiquid(p_52065_, p_52066_, p_52065_.m_8055_(p_52066_));
++   }
++
++   private static boolean touchesLiquid(BlockGetter p_52065_, BlockPos p_52066_, BlockState currentState) {
```
This added method overload should be declared in the exc file so that the parameters that are matching vanilla can be properly mapped when we have something that supports mapping parameters again.

pupnewfster commented on patches/minecraft/net/minecraft/world/level/block/ConcretePowderBlock.java.patch:
```diff
@@ -0,0 +1,39 @@
+       return m_52088_(p_52083_) || m_52064_(p_52081_, p_52082_);
+    }
+ 
++   // FORGE: Use BlockState overload
```
Add a deprecated annotation in front of the comment

pupnewfster commented on patches/minecraft/net/minecraft/world/level/block/FarmBlock.java.patch:
```diff
@@ -21,7 +21,12 @@
+       for(BlockPos blockpos : BlockPos.m_121940_(p_53260_.m_142082_(-4, 0, -4), p_53260_.m_142082_(4, 1, 4))) {
+-         if (p_53259_.m_6425_(blockpos).m_76153_(FluidTags.f_13131_)) {
++         net.minecraft.world.level.material.FluidState fluidState = p_53259_.m_6425_(blockpos);
++          if (fluidState.canHydrate(blockState)) {
```
Is there any reason this is assigning a local fluidstate variable instead of just acting on it directly since it is only used once below.

pupnewfster commented on patches/minecraft/net/minecraft/world/level/block/SugarCaneBlock.java.patch:
```diff
@@ -33,6 +33,15 @@
+             for(Direction direction : Direction.Plane.HORIZONTAL) {
+                BlockState blockstate1 = p_57176_.m_8055_(blockpos.m_142300_(direction));
+                FluidState fluidstate = p_57176_.m_6425_(blockpos.m_142300_(direction));
+-               if (fluidstate.m_76153_(FluidTags.f_13131_) || blockstate1.m_60713_(Blocks.f_50449_)) {
++               if (blockstate1.m_60713_(Blocks.f_50449_) || fluidstate.canHydrate(p_57175_)) {
```
This may be fine as is, but I wonder if it should be switched in order to match the order vanilla checks it just so it is a bit easier to see what the patch is doing even though comparing block type is likely less expensive of an operation than some can hydrate implementations may be.

pupnewfster commented on src/main/java/net/minecraftforge/client/IFluidRenderProperties.java:
```diff
@@ -0,0 +1,130 @@
+
+import javax.annotation.Nullable;
+
+public interface IFluidRenderProperties
```
I am going to guess not given I didn't see it anywhere but how do these methods handle being called if a modder were to call them on vanilla methods. Do they return proper values such as for the fog color? If not they probably should either be changed to do so or mention it in the javadocs.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java:
```diff
@@ -81,26 +87,41 @@ public Fluid getSource()
+        return false;
+    }
+
+    /**
+     * Used to determine if a {@link ForgeFlowingFluid} block can multiply and create new {@link Source} blocks.
+     *
+     * @param state The current {@link BlockState}.
+     * @param level The {@link Level} the formation is checked against.
+     * @param blockpos The {@link BlockPos} to determine if a {@link Source} should be created at.
+     * @return Returns whether a new {@link Source} block can be created.
+     */
+    @Override
+    protected FluidState getNewLiquid(LevelReader level, BlockPos blockpos, BlockState state)
+    {
+        return super.getNewLiquid(level, blockpos, state);
     }
```
Why is this method here just to call super and return the super value?

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/animal/Fox.java.patch:
```diff
@@ -27,3 +27,12 @@
+ 
+       public boolean m_8036_() {
+-         return Fox.this.m_20069_() && Fox.this.m_20120_(FluidTags.f_13131_) > 0.25D || Fox.this.m_20077_();
++         return Fox.this.m_20069_() && Fox.this.m_20120_(FluidTags.f_13131_) > 0.25D || Fox.this.m_20077_() || Fox.this.isInFluid() && Fox.this.getFluidHeight(Fox.this.getTouchingFluid()) > Fox.this.m_20204_();
```
Not sure if this should be mimic'd/if there is maybe even a mojang issue open about it, but I am unsure if this should be comparing against getFluidJumpThreshold (which is logical) or against the `0.25` that vanilla is using for in water.

Lanse505 replied:
My opinion is that we shouldn't change it and introduce new behaviour here.
We just "supplement" out behaviour on top if needed for this.

Lanse505 committed:
Fluid PR - Squashed

- IForgeFluid(State)
    - Add in node type handlers for pathfinding
    - Handle jump, motion, and acceleration of fluid movement
    - Add if the fluid can be swam in
- FluidAttributes
    - Add attribute for motion scaling of fluid pushing
- IForgeBlock(State)
    - Add in adjacent node type handlers for pathfinding
- IForgeEntity
    - Check to see whether the entity can be pushed by fluids and defaults to the water check
- WalkNodeProcessor/PathNodeType
   - Updated to accept new pathfinding methods within block and fluid
- FoxEntity/SwimTask/SwimGoal
   - Allow entities to swim in fluids
- ItemEntity
   - Allow items to float in fluids
- Entity
   - Added withinFluid/customEyeFluidLevelMap to handle custom fluid motion without tags
    - setInFluid/isInFluid
    - addEyeFluidLevel/getEyeFluidLevel
    - areEyesInCustomFluid
    - canTriggerSwimming
    - Patch isVisuallySwimming and updateFluidState for handling acceleration
- LivingEntity
    - Add fluid handler for motion and fluid jump
- Added Test Class

Fluid Animations, Interactions, and Farmland

- IForgeFluid(State)
    - Move swimming and drowning to attributes
    - Add fluid to fluid interaction
- IForgeLivingEntity
    - Add ability to check if entity can breathe in fluid
- ClientPlayerEntity
    - Patch animation handling for fluids
- Entity
    - Update swimming data and fix fluid bugs for swimming in no fluid
- LivingEntity
    - Handle fluid breathing
- Fluid Attributes
    - Add modifiers such as fall distance, can swim/drown/extinguish/hydrate
- Farmland
    - Add ability to hydrate farmland
- Add Test Cases

Block Interactions and Overlays

- IForgeFluid
    - Cleanup and bug fixes
- IForgeFluidState
    - Add hydration access
- FluidAttributes
    - Add hydration and view overlays
- Entity
    - Expose fluid getter
- LivingEntity
    - Fix fluid mechanics
- SpongeBlock/FarmlandBlock/ConcretePowderBlock
    - Add fluid interaction possibilities through hydration
- OverlayRenderer/FogRenderer/ForgeHooksClient
    - Add view overlays and fog color while in fluids
- Add test cases

Added JavaDocs and Changed Functions/BiFunctions to Predicate/BiPredicate where applicable for FluidAttributes.

Implement TriPredicate<FluidState, World, BlockPos> for canMultiply for Modded Fluids.
Apply Patch to needed area.

Clean-up Methods and Attributes.
Added JavaDocs.

- Killed off IFluidBlock
- Cleaned-Up JavaDocs in FluidUtil and removed useless check for IFluidBlock

- Fixed broken test mod

- Killed off FluidAttributes#isGaseous
- Added "forge:gaseous" tag

- Started work on splitting test mod into test mod>s<

Cleanup Patches and Javadocs

ForgeFlowingFluid
    - Fix recursive method call to super
IForgeFluid
    - Fix wrong eye level being stored causing all height checks to be true
IForgeEntity/LivingEntity
    - Add correct method checks for if the player is flying to allow flying movement through fluids
SwimGoal/SwimTask/FoxEntity
    - Add check to make sure that the eye is not above the fluid before swimming

Working Fluid Physicsish

- IForgeFluid(State)
    - Add in node type handlers for pathfinding
    - Handle jump, motion, and acceleration of fluid movement
    - Add if the fluid can be swam in
- FluidAttributes
    - Add attribute for motion scaling of fluid pushing
- IForgeBlock(State)
    - Add in adjacent node type handlers for pathfinding
- IForgeEntity
    - Check to see whether the entity can be pushed by fluids and defaults to the water check
- WalkNodeProcessor/PathNodeType
   - Updated to accept new pathfinding methods within block and fluid
- FoxEntity/SwimTask/SwimGoal
   - Allow entities to swim in fluids
- ItemEntity
   - Allow items to float in fluids
- Entity
   - Added withinFluid/customEyeFluidLevelMap to handle custom fluid motion without tags
    - setInFluid/isInFluid
    - addEyeFluidLevel/getEyeFluidLevel
    - areEyesInCustomFluid
    - canTriggerSwimming
    - Patch isVisuallySwimming and updateFluidState for handling acceleration
- LivingEntity
    - Add fluid handler for motion and fluid jump
- Added Test Class

Fluid Animations, Interactions, and Farmland

- IForgeFluid(State)
    - Move swimming and drowning to attributes
    - Add fluid to fluid interaction
- IForgeLivingEntity
    - Add ability to check if entity can breathe in fluid
- ClientPlayerEntity
    - Patch animation handling for fluids
- Entity
    - Update swimming data and fix fluid bugs for swimming in no fluid
- LivingEntity
    - Handle fluid breathing
- Fluid Attributes
    - Add modifiers such as fall distance, can swim/drown/extinguish/hydrate
- Farmland
    - Add ability to hydrate farmland
- Add Test Cases

Block Interactions and Overlays

- IForgeFluid
    - Cleanup and bug fixes
- IForgeFluidState
    - Add hydration access
- FluidAttributes
    - Add hydration and view overlays
- Entity
    - Expose fluid getter
- LivingEntity
    - Fix fluid mechanics
- SpongeBlock/FarmlandBlock/ConcretePowderBlock
    - Add fluid interaction possibilities through hydration
- OverlayRenderer/FogRenderer/ForgeHooksClient
    - Add view overlays and fog color while in fluids
- Add test cases

Added JavaDocs and Changed Functions/BiFunctions to Predicate/BiPredicate where applicable for FluidAttributes.

Implement TriPredicate<FluidState, World, BlockPos> for canMultiply for Modded Fluids.
Apply Patch to needed area.

Clean-up Methods and Attributes.
Added JavaDocs.

- Fixed broken test mod

- Killed off FluidAttributes#isGaseous
- Added "forge:gaseous" tag

- Started work on splitting test mod into test mod>s<

Cleanup Patches and Javadocs

ForgeFlowingFluid
    - Fix recursive method call to super
IForgeFluid
    - Fix wrong eye level being stored causing all height checks to be true
IForgeEntity/LivingEntity
    - Add correct method checks for if the player is flying to allow flying movement through fluids
SwimGoal/SwimTask/FoxEntity
    - Add check to make sure that the eye is not above the fluid before swimming

Killed off patches that got through rebase.


Start Port to 1.17


Fix fluid patches and maybe other things too

Cleaned-up Test Mod


Cleaned-up Patch documentation and Readded Coremod redirection call 


Reduced visibility to make transformer to work, cleaned-up comments


Fix test mod being enabled at all times!


Added Switches to the Test mods


Fixed broken test.


Tested Test mods after port.

FluidInteractionTest (Working)
FluidRenderTypeTest (Working)
FluidSourceTest (Working)

Fluid work for the day


Potential fog fix, needs testing

Potential fog fixes again

Fix fog and add boats

Work on Javadocs, also clean-up new methods.


Added patches for BoatDispenserBehaviour, SugarCaneBlock and CoralBlock


Javadocs clean-up


Some work before mini-weekation


Add render properties

Fix param name

clean-up


clean-up todos


Fix concrete powder fluid interactions

Some javadocs clean-up and fixed the last todo


Attempt to rebase part 1, push to check diffs

Add back missing patches

Apply missing Patch and revert build.gradle updates


started work on final clean-up for release


hopefully final tweaks before I can push


more clean-up


Moved methods that could be moved into IForgeEntity, fixed patch spacing


Rectify patch comments


Defaults should now "perfectly" mimic vanilla (I think)


Update FluidAttributes to match up closer to vanilla to avoid potential

desync issues due to datapack corruption.
Replaced vanillas checks for the default attribute checks.


Started work on comments


more work on comments

Lanse505 committed:
Fluid PR - Squashed

- IForgeFluid(State)
    - Add in node type handlers for pathfinding
    - Handle jump, motion, and acceleration of fluid movement
    - Add if the fluid can be swam in
- FluidAttributes
    - Add attribute for motion scaling of fluid pushing
- IForgeBlock(State)
    - Add in adjacent node type handlers for pathfinding
- IForgeEntity
    - Check to see whether the entity can be pushed by fluids and defaults to the water check
- WalkNodeProcessor/PathNodeType
   - Updated to accept new pathfinding methods within block and fluid
- FoxEntity/SwimTask/SwimGoal
   - Allow entities to swim in fluids
- ItemEntity
   - Allow items to float in fluids
- Entity
   - Added withinFluid/customEyeFluidLevelMap to handle custom fluid motion without tags
    - setInFluid/isInFluid
    - addEyeFluidLevel/getEyeFluidLevel
    - areEyesInCustomFluid
    - canTriggerSwimming
    - Patch isVisuallySwimming and updateFluidState for handling acceleration
- LivingEntity
    - Add fluid handler for motion and fluid jump
- Added Test Class

Fluid Animations, Interactions, and Farmland

- IForgeFluid(State)
    - Move swimming and drowning to attributes
    - Add fluid to fluid interaction
- IForgeLivingEntity
    - Add ability to check if entity can breathe in fluid
- ClientPlayerEntity
    - Patch animation handling for fluids
- Entity
    - Update swimming data and fix fluid bugs for swimming in no fluid
- LivingEntity
    - Handle fluid breathing
- Fluid Attributes
    - Add modifiers such as fall distance, can swim/drown/extinguish/hydrate
- Farmland
    - Add ability to hydrate farmland
- Add Test Cases

Block Interactions and Overlays

- IForgeFluid
    - Cleanup and bug fixes
- IForgeFluidState
    - Add hydration access
- FluidAttributes
    - Add hydration and view overlays
- Entity
    - Expose fluid getter
- LivingEntity
    - Fix fluid mechanics
- SpongeBlock/FarmlandBlock/ConcretePowderBlock
    - Add fluid interaction possibilities through hydration
- OverlayRenderer/FogRenderer/ForgeHooksClient
    - Add view overlays and fog color while in fluids
- Add test cases

Added JavaDocs and Changed Functions/BiFunctions to Predicate/BiPredicate where applicable for FluidAttributes.

Implement TriPredicate<FluidState, World, BlockPos> for canMultiply for Modded Fluids.
Apply Patch to needed area.

Clean-up Methods and Attributes.
Added JavaDocs.

- Killed off IFluidBlock
- Cleaned-Up JavaDocs in FluidUtil and removed useless check for IFluidBlock

- Fixed broken test mod

- Killed off FluidAttributes#isGaseous
- Added "forge:gaseous" tag

- Started work on splitting test mod into test mod>s<

Cleanup Patches and Javadocs

ForgeFlowingFluid
    - Fix recursive method call to super
IForgeFluid
    - Fix wrong eye level being stored causing all height checks to be true
IForgeEntity/LivingEntity
    - Add correct method checks for if the player is flying to allow flying movement through fluids
SwimGoal/SwimTask/FoxEntity
    - Add check to make sure that the eye is not above the fluid before swimming

Working Fluid Physicsish

- IForgeFluid(State)
    - Add in node type handlers for pathfinding
    - Handle jump, motion, and acceleration of fluid movement
    - Add if the fluid can be swam in
- FluidAttributes
    - Add attribute for motion scaling of fluid pushing
- IForgeBlock(State)
    - Add in adjacent node type handlers for pathfinding
- IForgeEntity
    - Check to see whether the entity can be pushed by fluids and defaults to the water check
- WalkNodeProcessor/PathNodeType
   - Updated to accept new pathfinding methods within block and fluid
- FoxEntity/SwimTask/SwimGoal
   - Allow entities to swim in fluids
- ItemEntity
   - Allow items to float in fluids
- Entity
   - Added withinFluid/customEyeFluidLevelMap to handle custom fluid motion without tags
    - setInFluid/isInFluid
    - addEyeFluidLevel/getEyeFluidLevel
    - areEyesInCustomFluid
    - canTriggerSwimming
    - Patch isVisuallySwimming and updateFluidState for handling acceleration
- LivingEntity
    - Add fluid handler for motion and fluid jump
- Added Test Class

Fluid Animations, Interactions, and Farmland

- IForgeFluid(State)
    - Move swimming and drowning to attributes
    - Add fluid to fluid interaction
- IForgeLivingEntity
    - Add ability to check if entity can breathe in fluid
- ClientPlayerEntity
    - Patch animation handling for fluids
- Entity
    - Update swimming data and fix fluid bugs for swimming in no fluid
- LivingEntity
    - Handle fluid breathing
- Fluid Attributes
    - Add modifiers such as fall distance, can swim/drown/extinguish/hydrate
- Farmland
    - Add ability to hydrate farmland
- Add Test Cases

Block Interactions and Overlays

- IForgeFluid
    - Cleanup and bug fixes
- IForgeFluidState
    - Add hydration access
- FluidAttributes
    - Add hydration and view overlays
- Entity
    - Expose fluid getter
- LivingEntity
    - Fix fluid mechanics
- SpongeBlock/FarmlandBlock/ConcretePowderBlock
    - Add fluid interaction possibilities through hydration
- OverlayRenderer/FogRenderer/ForgeHooksClient
    - Add view overlays and fog color while in fluids
- Add test cases

Added JavaDocs and Changed Functions/BiFunctions to Predicate/BiPredicate where applicable for FluidAttributes.

Implement TriPredicate<FluidState, World, BlockPos> for canMultiply for Modded Fluids.
Apply Patch to needed area.

Clean-up Methods and Attributes.
Added JavaDocs.

- Fixed broken test mod

- Killed off FluidAttributes#isGaseous
- Added "forge:gaseous" tag

- Started work on splitting test mod into test mod>s<

Cleanup Patches and Javadocs

ForgeFlowingFluid
    - Fix recursive method call to super
IForgeFluid
    - Fix wrong eye level being stored causing all height checks to be true
IForgeEntity/LivingEntity
    - Add correct method checks for if the player is flying to allow flying movement through fluids
SwimGoal/SwimTask/FoxEntity
    - Add check to make sure that the eye is not above the fluid before swimming

Killed off patches that got through rebase.

Start Port to 1.17

Fix fluid patches and maybe other things too

Cleaned-up Test Mod

Cleaned-up Patch documentation and Readded Coremod redirection call

Reduced visibility to make transformer to work, cleaned-up comments

Fix test mod being enabled at all times!

Added Switches to the Test mods

Fixed broken test.

Tested Test mods after port.

FluidInteractionTest (Working)
FluidRenderTypeTest (Working)
FluidSourceTest (Working)

Fluid work for the day

Potential fog fix, needs testing

Potential fog fixes again

Fix fog and add boats

Work on Javadocs, also clean-up new methods.

Added patches for BoatDispenserBehaviour, SugarCaneBlock and CoralBlock

Javadocs clean-up

Some work before mini-weekation

Add render properties

Fix param name

clean-up

clean-up todos

Fix concrete powder fluid interactions

Some javadocs clean-up and fixed the last todo

Attempt to rebase part 1, push to check diffs

Add back missing patches

Apply missing Patch and revert build.gradle updates

started work on final clean-up for release

hopefully final tweaks before I can push

more clean-up

Moved methods that could be moved into IForgeEntity, fixed patch spacing

Rectify patch comments

Defaults should now "perfectly" mimic vanilla (I think)

Update FluidAttributes to match up closer to vanilla to avoid potential

desync issues due to datapack corruption.
Replaced vanillas checks for the default attribute checks.

Started work on comments

more work on comments

pupnewfster commented on src/main/resources/forge.exc:
```diff
@@ -49,4 +49,5 @@ net/minecraft/world/level/block/FlowerPotBlock.<init>(Ljava/util/function/Suppli
-net/minecraft/world/level/storage/loot/LootPool.<init>([Lnet/minecraft/world/level/storage/loot/entries/LootPoolEntryContainer;[Lnet/minecraft/world/level/storage/loot/predicates/LootItemCondition;[Lnet/minecraft/world/level/storage/loot/functions/LootItemFunction;Lnet/minecraft/world/level/storage/loot/providers/number/NumberProvider;Lnet/minecraft/world/level/storage/loot/providers/number/Numbe ...
\ No newline at end of file
+net/minecraft/world/level/storage/loot/LootPool.<init>([Lnet/minecraft/world/level/storage/loot/entries/LootPoolEntryContainer;[Lnet/minecraft/world/level/storage/loot/predicates/LootItemCondition;[Lnet/minecraft/world/level/storage/loot/functions/LootItemFunction;Lnet/minecraft/world/level/storage/loot/providers/number/NumberProvider;Lnet/minecraft/world/level/storage/loot/providers/number/Numbe ...
+net/minecraft/world/level/block/ConcretePowderBlock.touchesLiquid(Lnet/minecraft/world/level/BlockGetter;Lnet/minecraft/core/BlockPos;Lnet/minecraft/world/level/block/state/BlockState;)V=|p_52065_,p_52066_,currentState
```
The addition of the concrete powder block should be able the loot pool one as `block` comes before `storage` when sorted alphabetically. If it hasn't been done the `checkExcs` task should be run, though I forget if that task enforces sort order and resorts or not.

pupnewfster commented on src/main/java/net/minecraftforge/common/data/ForgeFluidTagsProvider.java:
```diff
@@ -22,8 +22,7 @@
 import net.minecraftforge.common.ForgeMod;
-
-import static net.minecraftforge.common.Tags.Fluids.MILK;
+import net.minecraftforge.common.Tags.*;
```
This is not what I meant by my last review, I meant static import `Tags.Fluids.*` like the other ones static import `Tags.Items.*` etc

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java:
```diff
@@ -570,25 +570,66 @@ default boolean isBurning(BlockGetter world, BlockPos pos)
+     * <li>Negative Values = Untraversable</li>
+     * <li>0 = Best</li>
+     * <li>Highest = Worst</li>
+     * </ul>
```
What is this javadoc referencing as from what I can see neither the params nor the returns have numbers so can't be negative or things. The same goes for the variants of this method that are being added as they have the same javadoc

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java:
```diff
@@ -570,25 +570,66 @@ default boolean isBurning(BlockGetter world, BlockPos pos)
 
     /**
+     * TODO: Rename to {@code getPathNodeType} 1.17
+     * <br>
```
Leftover TODO?

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeFluid.java:
```diff
@@ -19,103 +19,333 @@
     {
-        return state.getExplosionResistance();
+        if (entity.touchingUnloadedChunk()) return false;
+        else
+        {
```
Personally I would say remove this else statement as the other part of it always returns so there is no real need to have all the contents of this else indented one spot further

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -380,6 +530,12 @@ public final Builder color(int color)
         public final Builder overlay(ResourceLocation texture)
         {
             overlayTexture = texture;
+            return viewOverlay(texture);
+        }
+
+        public final Builder viewOverlay(ResourceLocation texture)
+        {
+            viewOverlayTexture = new ResourceLocation(texture.getNamespace(), "textures/" + texture.getPath() + ".png");
```
Are there any cases where someone may wish to set an overlay but not the viewOverlay? (Not fully sure how the two overlays interact/work)

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -355,8 +496,17 @@ public static Builder builder(ResourceLocation stillTexture, ResourceLocation fl
+        private BiPredicate<FluidState, Entity> canExtinguish = (state, entity) -> state.is(FluidTags.WATER);
+        private BiPredicate<FluidState, BlockState> canHydrate = (fluidState, blockState) -> fluidState.is(FluidTags.WATER);
+        private BiFunction<FluidState, LivingEntity, Float> horizontalMotionModifier = (state, entity) -> entity.isSprinting() ? 0.9F : 0.8F;
+        private BiFunction<Enchantment, LivingEntity, Float> enchantmentModifier = (enchantment, entity) -> enchantment instanceof WaterWalkerEnchantment ? EnchantmentHelper.getDepthStrider(entity) : 0F;
```
I am not quite sure if this is the correct default for the enchantmentModifier or if this should be explicitly just set on forge's water fluid attributes as the default, as from what I can see of the default behavior vanilla has:
isInWater -> getDepthStrider, but this default would make someone seeing what enchantment modifiers lava has for example return it supports it instead of not. Alternatively if the overall consensus is that custom fluids should default to supporting depth strider, then potentially lava's fluid attribute should explicitly just return zero as the enchantment modifier for this.

Lanse505 committed:
Adress comments

Lanse505 committed:
Adress comments

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,355 @@
+            new TestOverrideFluid.Source(makeProperties())
+    );
+
+    public static RegistryObject<FlowingFluid> testOverRideFluidFlowing = FLUIDS.register("test_override_fluid_flowing", () ->
```
Based on the other variable names I think this one has the wrong capitalization.

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,355 @@
+        }
+    }
+
+    private static class FluidloggableBlock extends Block implements SimpleWaterloggedBlock
```
I would say given this is an example for people this really should implement `BucketPickup` and `LiquidBlockContainer` instead of `SimpleWaterloggedBlock` as it doesn't actually support water and all the methods except `getPickupSound` are being overridden anyways. That way if anyone copies this style of doing it into their mod they don't run into any issues if another mod is doing an instanceof check on SimpleWaterloggedBlock (I vaguely recall running into at least one mod doing that at some point)

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,355 @@
+                }
+                return true;
+            }
+            else
```
Might as well remove the else from here and just return false so it is more readable as it only takes one line instead of four.

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,355 @@
+                world.setBlock(pos, blockState.setValue(FLUIDLOGGED, false), 3);
+                return new ItemStack(blockState.getFluidState().getType().getBucket());
+            }
+            else
```
Might as well remove the else from here and just return false so it is more readable as it only takes one line instead of four.

Lanse505 committed:
Adress new comments

Lanse505 committed:
Adress new comments

Lanse505 committed:
Fix accidental exc removal

Lanse505 committed:
Fix accidental exc removal

rehwinkel commented on patches/minecraft/net/minecraft/world/level/block/LiquidBlock.java.patch:
```diff
@@ -1,10 +1,16 @@
     public static final ImmutableList<Direction> f_181233_ = ImmutableList.of(Direction.DOWN, Direction.SOUTH, Direction.NORTH, Direction.EAST, Direction.WEST);
  
-+   @Deprecated  // Forge: Use the constructor that takes a supplier
++   @Deprecated  // FORGE: Use the constructor that takes a supplier
```
Is this change useful?

Lanse505 replied:
I can revert it, I changed it to match the other of the Deprecation comment in the class.
It's a arbitrary change, and I can change it back if the forge team says so.

KingLemming commented:
One general comment I have (and really, I will look over it more as I can) - I thought we were getting rid of viscosity as an attribute? It still seems to be there.

Lanse505 commented:
> 
> 
> One general comment I have (and really, I will look over it more as I can) - I thought we were getting rid of viscosity as an attribute? It still seems to be there.

We agreed awhile back to not remove some of these fluid attributes in this PR.
But rather go back at a later date in a seperate PR and kill off un-used attributes and methods, and do general clean-up.

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+        @Override
+        public boolean canPlaceLiquid(BlockGetter world, BlockPos pos, BlockState state, Fluid fluid)
+        {
+            return !state.getValue(FLUIDLOGGED) && (state.getFluidState().isEmpty() || state.getFluidState().getType() == fluid);
```
I don't think this is quite right.. when I was doing a bit of testing using a water bucket on this block or placing it in water causes it to become fluid logged with the custom fluid. So given this is an example this should probably include a check similar to how vanilla's SimpleWaterloggedBlock is that it validates that `fluid` is the type of fluid that it can be fluid logged with.

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+        }
+    }
+
+    private static class FluidloggableBlock extends Block implements BucketPickup, LiquidBlockContainer
```
It probably doesn't matter given it is just a test mod, but one thing that surprised me for a little until I remembered why it is the case... is that the fluid doesn't flow out of this block when it is fluid logged. Adding a non full block shape would "fix" this (potentially just the empty voxelshape given no collision is set in the block's properties)

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+        }
+
+        @Override
+        public Optional<SoundEvent> getPickupSound() {
```
Bracket should be on new line

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+                entity.setDeltaMovement(entity.getDeltaMovement().add(0.0D, (double)-0.1F * entity.getAttribute(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).getValue(), 0.0D));
+            }
+
+            // If the custom fluid is trying to place a flowing block, and it touches a Lava block.
```
This should probably say touches a Lava source block (or the code below changed to check the lava tag)

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+            }
+
+            @Override
+            public void sink(FluidState state, LivingEntity entity) {
```
Bracket should be on new line

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+                entity.setDeltaMovement(entity.getDeltaMovement().add(0.0D, (double)-0.1F * entity.getAttribute(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).getValue(), 0.0D));
+            }
+
+            // If the custom fluid is trying to place a flowing block, and it touches a Lava block.
```
This should probably say touches a Lava source block (or the code below changed to check the lava tag)

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+            }
+
+            @Override
+            public void sink(FluidState state, LivingEntity entity) {
```
Bracket should be on new line

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+
+    public CoreFluidTest()
+    {
+        if (ENABLE) {
```
Bracket should be on new line

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+import java.util.stream.Stream;
+
+@Mod(CoreFluidTest.MODID)
+public class CoreFluidTest {
```
Bracket should be on new line

pupnewfster commented on src/test/java/net/minecraftforge/debug/client/rendering/ShaderFixTest.java:
```diff
@@ -31,9 +31,13 @@
     public static void clientSetup(final FMLClientSetupEvent event)
     {
-        ClientRegistry.registerEntityShader(LocalPlayer.class, new ResourceLocation("shaders/post/desaturate.json"));
+        if (ENABLED) {
```
Bracket should be on new line

ChampionAsh5357 replied:
I mean out of scope for this PR but sure.

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,357 @@
+package net.minecraftforge.debug.fluid;
```
Missing license header

pupnewfster commented on src/main/java/net/minecraftforge/client/IFluidRenderProperties.java:
```diff
@@ -0,0 +1,134 @@
+    default void setupFog(Camera camera, FogRenderer.FogMode fogMode, float renderDistance)
+    {
+        float constant = 180.0F;
+        if (camera.getEntity() instanceof LocalPlayer player) {
```
Bracket should be on new line

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeLivingEntity.java:
```diff
@@ -0,0 +1,23 @@
+package net.minecraftforge.common.extensions;
```
Missing license header

pupnewfster commented on src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java:
```diff
@@ -48,12 +50,16 @@
+    /**
+     * The {@link TriPredicate} used to determine if the {@link Fluid} can multiply and create new source blocks.
+     */
+    private final TriPredicate<FluidState, LevelReader, BlockPos> canMultiply;
```
This field seems to be unused?

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeEntity.java:
```diff
@@ -197,4 +199,79 @@ default boolean isMultipartEntity()
+     */
+    default boolean isPushedByFluid(FluidState state)
+    {
+        return !(self() instanceof Player) || !((Player) self()).getAbilities().flying;
```
This might be a little cleaner using pattern matching: `return !(self() instanceof Player player) || !player.getAbilities().flying;`

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeEntity.java:
```diff
@@ -197,4 +199,79 @@ default boolean isMultipartEntity()
+     */
+    default boolean isInFluid()
+    {
+        return self().getTouchingFluid() != net.minecraft.world.level.material.Fluids.EMPTY.defaultFluidState() && self().getFluidHeight(self().getTouchingFluid()) > 0.0D;
```
Is there a reason this is using the fully qualified name for checking against the empty fluid rather than just importing it?

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeEntity.java:
```diff
@@ -197,4 +199,79 @@ default boolean isMultipartEntity()
+    }
+
+    /**
+     * Used to check if the entities eyes are inside of a "custom" fluid, aka a fluid that isn't added by vanilla.
```
I believe this should be `entity's eyes` not `entities eyes`

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeFluid.java:
```diff
@@ -19,103 +19,334 @@
-    FluidAttributes getAttributes();
+    default void sink(FluidState state, LivingEntity entity)
+    {
+        entity.setDeltaMovement(entity.getDeltaMovement().add(0.0D, (double)-0.04F * entity.getAttribute(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).getValue(), 0.0D));
```
Why is ForgeMod.SWIM_SPEED referenced by the fully qualified name here when ForgeMod is already imported.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -229,20 +300,27 @@ public Component getDisplayName(FluidStack stack)
 
     /**
      * A FluidStack sensitive version of getTranslationKey
+     * @param stack the provided {@code FluidStack} to get the display name of.
```
I think this should be to get the translation key of not display name of?

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -200,27 +269,29 @@ public boolean doesVaporize(BlockAndTintGetter reader, BlockPos pos, FluidStack
+     * @param stack the provided {@code FluidStack} to get the display name of.
+     * @return Returns the display name of the provided {@code FluidStack} as a {@code Component}.
      */
     public Component getDisplayName(FluidStack stack)
```
The body of this method (while not touched by this PR) should be fixed to pass the stack to the `getTranslationKey` method being used.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -229,20 +300,27 @@ public Component getDisplayName(FluidStack stack)
-     * Returns the translation key of this fluid.
+     * @return Returns the translation key of this fluid.
      */
     public String getTranslationKey()
```
I feel like this method potentially should be deprecated/slated to be removed for 1.18 as I can't see much reason for it given there is a more sensitive version and if a mod overrides one but not the other then things may behave unexpectedly.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -229,20 +300,27 @@ public Component getDisplayName(FluidStack stack)
         return this.translationKey;
     }
 
+    /**
```
I feel like this javadoc should be below the section comment of default accessors instead of above.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -279,6 +352,41 @@ public int getColor()
         return color;
     }
 
+    public double getMotionScale(FluidState state, Entity entity)
+    {
+        return motionScale.applyAsDouble(state, entity);
+    }
+
+    public double getFallDistanceModifier(FluidState state, Entity entity)
+    {
+        return fallDistanceModifier.applyAsDouble(state, entity);
+    }
+
+    public boolean canSwim(FluidState state)
+    {
+        return canSwim.test(state);
+    }
+
+    public boolean canDrown(FluidState state, LivingEntity entity)
+    {
+        return canDrown.test(state, entity);
+    }
+
+    public boolean canExtinguish(FluidState state, Entity entity)
+    {
+        return canExtinguish.test(state, entity);
+    }
+
+    public boolean canHydrate(FluidState fluidState, BlockState blockState)
+    {
+        return canHydrate.test(fluidState, blockState);
+    }
+
+    public boolean canBoat(FluidState state, Boat boat)
+    {
+        return this.canBoat.test(state, boat);
+    }
```
Part of me feels like it may be useful to add the javadocs to these getters, especially given they are mostly written already on the actual fields. That way it is easier to see the javadocs of what these methods do when using the IDE rather than having to go to implementation and then look at the field and see that it happens to have a javadoc.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -432,6 +553,48 @@ public final Builder sound(SoundEvent fillSound, SoundEvent emptySound)
+        {
+            this.motionScale = motionScale;
+            return this;
+        }
+
+        public final Builder fallDistanceModifier(ToDoubleBiFunction<FluidState, Entity> fallDistanceModifier)
+        {
+            this.fallDistanceModifier = fallDistanceModifier;
+            return this;
+        }
+
+        public final Builder canSwim(Predicate<FluidState> canSwim)
+        {
+            this.canSwim = canSwim;
+            return this;
+        }
+
+        public final Builder canDrown(BiPredicate<FluidState, LivingEntity> canDrown)
+        {
+            this.canDrown = canDrown;
+            return this;
+        }
+
+        public final Builder canExtinguish(BiPredicate<FluidState, Entity> canExtinguish)
+        {
+            this.canExtinguish = canExtinguish;
+            return this;
+        }
+
+        public final Builder canHydrate(BiPredicate<FluidState, BlockState> canHydrate)
+        {
+            this.canHydrate = canHydrate;
+            return this;
+        }
+
+        public final Builder canBoat(BiPredicate<FluidState, Boat> canBoat)
+        {
+            this.canBoat = canBoat;
+            return this;
+        }
```
Given there are already javadocs for what these params are for, it may be helpful to also add them on these builder methods as the brief description of what it is for.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -446,13 +609,14 @@ protected Water(Builder builder, Fluid fluid)
 
         public static Builder builder(ResourceLocation stillTexture, ResourceLocation flowingTexture) {
-            return new Builder(stillTexture, flowingTexture, Water::new);
+            return new Builder(stillTexture, flowingTexture, Water::new).canExtinguish((state, entity) -> true);
```
Is there a reason this canExtinguish -> true  is not in ForgeHooks#createVanillaFluidAttributes where all the other water specific properties are set?

pupnewfster reviewed (commented):
I added a bunch more comments but in addition to those, there are a few things (I haven't tested all the features yet) in the test mod that don't seem to be working quite right:
- I can't swim in either the attribute or override fluids
- The override fluid isn't turning concrete powder into concrete (the attribute fluid is)
- The lighting of the test fluids is very different between still and flowing (may be intended/due to how the textures are not sure, but at least looks weird)

pupnewfster commented on patches/minecraft/net/minecraft/world/level/material/Fluid.java.patch:
```diff
@@ -35,6 +42,31 @@
++   }
++
++   private void initClient() {
++      if (net.minecraftforge.fml.loading.FMLEnvironment.dist == net.minecraftforge.api.distmarker.Dist.CLIENT) {
```
Should be updated to match https://github.com/MinecraftForge/MinecraftForge/commit/347a0b3c0dc7873b9bf6910f6ef9b99a2d786668

desht commented:
Would it be out of scope to talk about more first-class fluidstack NBT support in this PR?  In particular:

* Allow fluids & fluidstacks to provide capabilities in the same way items & itemstacks do
* Preserve fluidstack NBT when the fluid is bucketed (i.e. store fluidstack NBT in a forge-namespaced NBT field on the bucket item, and pay attention to it when fluid is later extracted from the bucket via `IFluidHandlerItem`)
* Preserve fluidstack NBT when the fluid is poured into the world, but if & only if the fluid block has a tile entity (so this is more of a niche case)
* An equivalent to `Item#appendHoverText()`; something like `IForgeFluid#appendHoverText(FluidStack, World, List<ITextComponent>, ITooltipFlag`.  Basically a standard way for mod authors to get fluidstack-specific tooltip data when rendering fluids in their GUI's (and probably only useful for custom information if the fluidstack has NBT data).

Lanse505 commented:
> 
> 
> Would it be out of scope to talk about more first-class fluidstack NBT support in this PR? In particular:
> 
>     * Allow fluids & fluidstacks to provide capabilities in the same way items & itemstacks do
> 
>     * Preserve fluidstack NBT when the fluid is bucketed (i.e. store fluidstack NBT in a forge-namespaced NBT field on the bucket item, and pay attention to it when fluid is later extracted from the bucket via `IFluidHandlerItem`)
> 
>     * Preserve fluidstack NBT when the fluid is poured into the world, but if & only if the fluid block has a tile entity (so this is more of a niche case)
> 
>     * An equivalent to `Item#appendHoverText()`; something like `IForgeFluid#appendHoverText(FluidStack, World, List<ITextComponent>, ITooltipFlag`.  Basically a standard way for mod authors to get fluidstack-specific tooltip data when rendering fluids in their GUI's (and probably only useful for custom information if the fluidstack has NBT data).

That is indeed out-of-scope for this particular PR

ChampionAsh5357 committed:
Address pup's comments and rebase

ChampionAsh5357 committed:
Actually gen the patches

ChampionAsh5357 committed:
Merge remote-tracking branch 'origin/fluid-physics-1.17' into fluid-physics-1.17

# Conflicts:
#	patches/minecraft/net/minecraft/world/level/block/ConcretePowderBlock.java.patch
#	patches/minecraft/net/minecraft/world/level/material/Fluid.java.patch
#	src/main/java/net/minecraftforge/client/IFluidRenderProperties.java
#	src/main/java/net/minecraftforge/common/extensions/IForgeEntity.java
#	src/main/java/net/minecraftforge/common/extensions/IForgeFluid.java
#	src/main/java/net/minecraftforge/common/extensions/IForgeFluidState.java
#	src/main/java/net/minecraftforge/common/extensions/IForgeLivingEntity.java
#	src/main/java/net/minecraftforge/fluids/FluidAttributes.java
#	src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java
#	src/test/java/net/minecraftforge/debug/client/rendering/ShaderFixTest.java
#	src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java
#	src/test/resources/META-INF/mods.toml

ChampionAsh5357 committed:
Add `canMultiply` and add test cases

ChampionAsh5357 committed:
Merge branch '1.17.x' into fluid-physics-1.17

ChampionAsh5357 committed:
Fix misapplied swimming patch

Fixes slow swimming in water

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,390 @@
+        }
+
+        @Override
+        public VoxelShape getShape(BlockState p_60555_, BlockGetter p_60556_, BlockPos p_60557_, CollisionContext p_60558_) {
```
Bracket should be on new line, and the parameters should have some proper names rather than srg names given this is in forge code.

pupnewfster commented on src/test/java/net/minecraftforge/debug/fluid/CoreFluidTest.java:
```diff
@@ -0,0 +1,390 @@
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.entity.vehicle.Boat;
+import net.minecraft.world.item.*;
```
Forge tries to stay away from wildcard imports (there are a few in this class)

pupnewfster commented on patches/minecraft/net/minecraft/world/level/material/FlowingFluid.java.patch:
```diff
@@ -18,3 +27,22 @@
++   @Override
++   public boolean canMultiply(FluidState state, LevelReader reader, BlockPos pos) {
++      return this.m_6760_();
+    }
```
The patch complexity of this file would be slightly less if this method implementation being patched in was added just above the deprecation of the method it is replacing so that it is all one hunk of:
```java
@Override
public boolean canMultiply(...) {
...
}

@Deprecated // Forge: Use canMultiply instead
```

pupnewfster reviewed (commented):
This looks pretty good now and seems to be working as expected, just a couple tiny formatting changes remaining (and it seems one was missed from the previous set of comments for the first bracket of the CoreFluidTest class).

The only "big" thing I am a bit unsure about currently, is that given how far into 1.17's life cycle we are and the fact we are trying to minimize breaking changes, I feel like it may not make sense at this point to be including the rename of `getAiPathNodeType` to `getBlockPathType` until 1.18 and just leave a TODO in the various places for the purposes of renaming it.

ChampionAsh5357 committed:
Address pup's comments and revert method renames

ChampionAsh5357 committed:
Merge branch '1.17.x' into fluid-physics-1.17

ChampionAsh5357 committed:
Fix formatting and readd `isGaseous` to reduce breaking changes, will be removed in 1.18

ShroomAgent27 commented:
I wrote a [basic mod](https://github.com/FoalFreePress/TestingForgeMods/tree/3c045facc1b48f51785051fcd243318b486d930b/7992/src/main/java/com/example/examplemod) that adds 8 fluids, and they all work as expected. But my mod also doesn't do anything fancy with them. They're just simple fluids.

pupnewfster commented on patches/minecraft/net/minecraft/world/level/block/CoralBlock.java.patch:
```diff
@@ -0,0 +1,11 @@
+       for(Direction direction : Direction.values()) {
+          FluidState fluidstate = p_52135_.m_6425_(p_52136_.m_142300_(direction));
+-         if (fluidstate.m_76153_(FluidTags.f_13131_)) {
++         if (fluidstate.canHydrate(p_52135_.m_8055_(p_52136_))) {
```
This blockstate should be cached similarly to how it is done in the FarmBlock so as to not have to look it up each time.

pupnewfster commented on patches/minecraft/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java.patch:
```diff
@@ -1,5 +1,21 @@
+ 
+-                  if (p_77608_.m_6425_(p_77609_).m_76153_(FluidTags.f_13131_)) {
++                  BlockPathTypes blockType = blockstate.getAdjacentBlockPathType(p_77608_, p_77609_, p_77610_);
++                  if (blockType != null) return blockType;
```
The above checks for cactus, sweet berries, and burning should be moved into the default implementation so if modders call the method directly they get the correct result. Similarly the `FluidState#getAdjacentBlockPathTypes` should have the check for water in its default.

Potentially I think it may even make sense to call the fluidstate one from the blockstate one given blockstates are currently strongly tied with their fluidstate.

pupnewfster commented on patches/minecraft/net/minecraft/client/player/LocalPlayer.java.patch:
```diff
@@ -28,7 +28,35 @@
+       if (this.m_20142_()) {
+          boolean flag5 = !this.f_108618_.m_108577_() || !flag4;
+-         boolean flag6 = flag5 || this.f_19862_ || this.m_20069_() && !this.m_5842_();
++         boolean flag6 = flag5 || this.f_19862_ || this.m_20069_() && !this.m_5842_() || this.isInFluid() && !this.canTriggerSwimming();
```
Add some clarifying parenthesis so it is easier to see where and how the ANDs and ORs interact

pupnewfster commented on patches/minecraft/net/minecraft/world/level/block/ConcretePowderBlock.java.patch:
```diff
@@ -0,0 +1,52 @@
+    private static boolean m_52088_(BlockState p_52089_) {
+-      return p_52089_.m_60819_().m_76153_(FluidTags.f_13131_);
++      return p_52089_.m_60819_().canHydrate(p_52089_);
+    }
```
This helper doesn't seem like it works quite properly as it is passing the block's state itself as the target state when instead of an actual target? Unless this is meant to be for seeing if a fluid logged block can be hydrated by itself??? In which case this could potentially use a comment.

Though it looks like most places that call this method have `|| fluidstate.canHydrate(state)` added, which seems like that call could replace the reference to this method all together?

ChampionAsh5357 replied:
Yeah, vanilla does the same exact thing where it checks whether the fluid state of the block is water which deals with some very specific cases when it comes to how fluids interact with so we need to check both the state of the block and adjacent block.

pupnewfster replied:
Ok, best bet then is to add a comment as well about it intentionally checking the same block.

pupnewfster commented on patches/minecraft/net/minecraft/world/level/material/FlowingFluid.java.patch:
```diff
@@ -18,3 +27,22 @@
++
++   @Override
++   public boolean canMultiply(FluidState state, LevelReader reader, BlockPos pos) {
++      return this.m_6760_();
```
If this is moved above the deprecation of vanilla's `canConvertToSource` that would make it be one hunk instead of two and thus a bit smaller of a patch. Also what is the reasoning for using the name `canMultiply` instead of `canConvertToSource`? It seems like `canMultiply` is most likely a remnant from MCP naming?

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -116,6 +141,8 @@
      * Generally this is associated with negative density fluids.
      */
+    // Check against {@link net.minecraftforge.common.Tags.Fluids#GASEOUS} instead.
+    @Deprecated(forRemoval = true, since = "1.18")
```
This and the other deprecations for removal in this class should be labeled using `since = "1.17.1"` as the decision at some point on how to use it got changed.

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeFluidState.java:
```diff
@@ -36,44 +45,215 @@ private FluidState self()
+    default BlockPathTypes getAdjacentBlockPathType(BlockGetter level, BlockPos pos, BlockPathTypes originalType)
+    {
+        return getAdjacentBlockPathType(level, pos, null, originalType);
+    }
```
This method should be removed and just have the caller pass in null to the version that takes an entity, as it is clearer than having an overload that has the extra parameter that got added in the middle in the middle.

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeFluid.java:
```diff
@@ -19,103 +19,333 @@
+     * @param boat The supplied {@link Boat} entity
+     * @return Whether the fluid supports boats being used with it.
+     */
+    default boolean canBoat(FluidState state, Boat boat)
```
This method (and other references to `canBoat`) need a better name. Potentially `supportsBoat` or `supportsBoating` would be better and make more sense?

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeFluidState.java:
```diff
@@ -36,44 +45,215 @@ private FluidState self()
+     * @param otherState The secondary state to check
+     * @return Whether the two provided {@link FluidState}s {@link Fluid}s match.
+     */
+    default boolean is(FluidState otherState)
```
This should probably either be renamed to `sameType` or if the reason the naming for it being `is` is because that is what vanilla calls some helpers in block states, then this should take the fluid type instead of the fluid state of the other fluid similar to how block states take the block to compare against and not the block state.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/item/ItemEntity.java.patch:
```diff
@@ -27,6 +27,15 @@
+          Vec3 vec3 = this.m_20184_();
+          float f = this.m_20192_() - 0.11111111F;
+-         if (this.m_20069_() && this.m_20120_(FluidTags.f_13131_) > (double)f) {
++         if (this.m_20069_() && this.m_20120_(FluidTags.f_13131_) > (double)f || this.isInFluid() && this.getFluidHeight(this.getTouchingFluid()) > f) {
```
Logically this should cause the else statement after this for lava to never get fired, as lava is a fluid so the added case after the OR will be true, though from looking at usage *unless* a modder specifically calls `setInFluid` or `updateFluidHeightAndDoFluidPushing` on lava/water it is only actually adjusted for custom fluids. This makes me wonder if this is actually the case if it should be `setInCustomFluid` and `isInCustomFluid` to make it clearer that it is not true for water/lava (and then maybe even add in some sort of validation to make sure it doesn't get called with water or lava?).

Though on second thought, is there a reason that this is an `||` instead of its own branch that then allows for a custom under fluid push amount to be applied in case mods want to add a value like lava has instead of only being able to use the one that in water has?

pupnewfster commented on src/main/java/net/minecraftforge/common/extensions/IForgeFluid.java:
```diff
@@ -19,103 +19,333 @@
      */
-    /* TODO: reimplement
-    default boolean canRenderInLayer(IFluidState state, BlockRenderLayer layer)
+    default void handleMotion(FluidState state, LivingEntity entity, Vec3 travelVector, double gravity)
```
Rather than copying a massive chunk of vanilla code for the default implementation of this, it would be much more manageable if this basically acted/was:
```java
//true -> same handling as water, false -> custom handling implemented as part of this method
default void handleMotion(FluidState state, LivingEntity entity, Vec3 travelVector, double gravity)
{
    return true;
}
```
and then in `LivingEntity#travel` instead of adding another if branch just doing this:
```java
- if (this.isInWater() && this.isAffectedByFluids() && !this.canStandOnFluid(fluidstate.getType())) {
+ if ((this.isInWater() || this.isInFluid()) && this.isAffectedByFluids() && !this.canStandOnFluid(fluidstate.getType())) {
+    if (this.isInWater() || this.getTouchingFluid().handleMotion(this, p_21280_, d0)) {
   //Vanilla water impl here
+    }
}
```

ChampionAsh5357 committed:
Address comments with clarifications, additions, and simplifications

ChampionAsh5357 committed:
Merge branch '1.17.x' into fluid-physics-1.17

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -365,24 +365,41 @@
              }
  
-+            f6 *= (float)this.m_21051_(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).m_22135_();
++            f6 *= (float) this.m_21051_(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).m_22135_();
```
Remove this spacing change so that this isn't part of the patch.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -365,24 +365,41 @@
+             float f5 = this.m_20142_() ? 0.9F : this.m_6108_();
+             float f6 = 0.02F;
+-            float f7 = (float)EnchantmentHelper.m_44922_(this);
++            float f7 = (float) EnchantmentHelper.m_44922_(this);
```
Remove this spacing change so that this isn't part of the patch.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -365,24 +365,41 @@
+             }
+ 
+-            this.m_20256_(vec36.m_82542_((double)f5, (double)0.8F, (double)f5));
++            this.m_20256_(vec36.m_82542_((double) f5, (double) 0.8F, (double) f5));
```
Remove this spacing change so that this isn't part of the patch.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -365,24 +365,41 @@
+-            if (this.f_19862_ && this.m_20229_(vec32.f_82479_, vec32.f_82480_ + (double)0.6F - this.m_20186_() + d8, vec32.f_82481_)) {
+-               this.m_20334_(vec32.f_82479_, (double)0.3F, vec32.f_82481_);
++            if (this.f_19862_ && this.m_20229_(vec32.f_82479_, vec32.f_82480_ + (double) 0.6F - this.m_20186_() + d8, vec32.f_82481_)) {
++               this.m_20334_(vec32.f_82479_, (double) 0.3F, vec32.f_82481_);
```
Remove this spacing change so that this isn't part of the patch.

ChampionAsh5357 committed:
Fix patches before working on performant tag physics

ChampionAsh5357 committed:
Merge remote-tracking branch 'origin/fluid-physics-1.17' into fluid-physics-1.17

pupnewfster commented on patches/minecraft/net/minecraft/client/player/LocalPlayer.java.patch:
```diff
@@ -37,6 +65,15 @@
+-      if (this.m_20069_() && this.f_108618_.f_108573_ && this.m_6129_()) {
+-         this.m_21208_();
++      if ((this.m_20069_() || this.isInCustomFluid()) && this.f_108618_.f_108573_ && this.m_6129_()) {
++         this.getTouchingCustomFluid().sink(this);
```
Given water is not a custom fluid I believe this breaks vanilla functionality.

Lanse505 replied:
Unless it is a custom fluid with overriden behaviour, it'll use the same "sinking" logic as vanilla.

Original: this.goDownInWater
```java
protected void goDownInWater() {
      this.setDeltaMovement(this.getDeltaMovement().add(0.0D, (double)-0.04F * this.getAttribute(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).getValue(), 0.0D));
   }
   ```
   
   Patch: this.getTouchingCustomFluid().sink(this)
   ```java
       default void sink(LivingEntity entity)
    {
        self().getType().sink(self(), entity);
    }
    
    Which calls
    
        default void sink(FluidState state, LivingEntity entity)
    {
        entity.setDeltaMovement(entity.getDeltaMovement().add(0.0D, (double)-0.04F * entity.getAttribute(ForgeMod.SWIM_SPEED.get()).getValue(), 0.0D));
    }
    ```

Lanse505 committed:
Merge remote-tracking branch 'MinecraftForge/1.17.x' into fluid-physics-1.17

Lanse505 committed:
Fix patch as to not break vanilla interactions

pupnewfster commented on patches/minecraft/net/minecraft/client/player/LocalPlayer.java.patch:
```diff
@@ -37,6 +65,15 @@
+       this.f_108614_ = this.m_21255_();
+       if (this.m_20069_() && this.f_108618_.f_108573_ && this.m_6129_()) {
+          this.m_21208_();
+-      }
++      } else if (this.isInCustomFluid() && this.f_108618_.f_108573_ && this.m_6129_()) this.getTouchingCustomFluid().sink(this);
```
The else if should just be on the next line so as to not have the patch of removing a bracket.

Lanse505 committed:
Made suggested change

sciwhiz12 commented:
Unfortunately, as the 1.18.1 RB draws nearer and the PR authors (@Lanse505 and @ChampionAsh5357) are currently busy with personal matters to update this PR to 1.18.x, we've had to make the decision to push this PR to 1.19.x.

If there is anyone willing to help out in updating this PR and responding to feedback, please contact the PR authors and disscuss the matter in the Forge project discord. We hope that, by the time 1.19 rolls around, this PR is updated quickly enough to get it in relatively early in the breaking changes window for 1.19 (avoiding breaking many mods with the fluids changes, as would likely happen if this were merged in 1.18).

For now, this PR has been marked as a draft to reflect its work-in-progress status. Thank you!

MrScautHD commented:
1.19 sad

Lanse505 commented:
Superseded by https://github.com/MinecraftForge/MinecraftForge/pull/8695
